### ISSUE TITLE
Unable to read Intel RAPL files for CPU power : Permission denied

### ISSUE THREAD

**Issue by benoit-cty** [mlco2/codecarbon#244]
## Description

Due to a [security issue](https://www.suse.com/security/cve/CVE-2020-8694.html) all Linux distributions are changing right permission of the RAPL file, used by Code Carbon for CPU power. After updating the file could  be read only by root.

There is a workaround, thanks to https://github.com/prometheus/node_exporter/issues/1892 :
create/modify `/etc/sysfs.conf` and reboot after adding:
`mode class/powercap/intel-rapl:0/energy_uj = 0444`

But you need root access to do this. Following https://github.com/mlco2/codecarbon/issues/232 that has been closed by a PR that reduce errors logs, I create this issue to find a workaround that don't need to be root.

We could use the CPU load and apply it to the CPU TDP, but it is not so simple as shown here https://medium.com/teads-engineering/estimating-aws-ec2-instances-power-consumption-c9745e347959

**Comment by Erotemic** [comment]
What is the security implication of doing this? In the security summary it says "information disclosure". 

Letting anyone read the file will show what? How much energy the CPU is using, and that could be used to infer that the computer is doing work? Is it any more serious than that?

**Comment by benoit-cty** [comment]
Yes it's just that, _the researchers behind PLATYPUS found that the RAPL interface reported power consumption with enough granularity to deduce vital secrets_ :astonished:  https://www.securityweek.com/platypus-hackers-can-obtain-crypto-keys-monitoring-cpu-power-consumption
Instead of limiting the access to root they could have reduced the granularity, but it's more difficult :cry: 
Code Carbon only look every few seconds for consumption, so the data collected are not enough to do this attack.

**Comment by jlaumonier** [comment]
Maybe a little improvement of the workaround proposed here to avoid giving access to all to `energy_uj` but with more steps :

1.    `sudo addgroup power`
2.    `sudo apt install sysfsutils`
3.    Add in `/etc/sysfs.conf`: 
```
       mode class/powercap/intel-rapl:0/energy_uj = 0440
       owner class/powercap/intel-rapl:0/energy_uj = root:power
```
 
4. add the new group `power` to the user who will use codecarbon
5. reboot

Other solution : someone seems to obtain the same results with systemd stuff : https://serverfault.com/questions/1123964/allow-a-systemd-service-access-to-sys-energy-uj-udev-rule-changing-owner-ha

**Comment by stefanovrizzi** [comment]
@benoit-cty, when I run on Linux:
```echo "mode class/powercap/intel-rapl:0/energy_uj = 0440" >> /etc/sysfs.conf```

I get:
```bash: /etc/sysfs.conf: Permission denied```

Same for:
```echo "owner class/powercap/intel-rapl:0/energy_uj = root:codecarbon" >> /etc/sysfs.conf```

May I please have some help?

> ## Description
> 
> Due to a [security issue](https://www.cve.org/CVERecord?id=CVE-2020-8694) all Linux distributions are changing right permission of the RAPL file, used by Code Carbon for CPU power. After updating the file could be read only by root.
> 
> There is a workaround, thanks to [prometheus/node_exporter#1892](https://github.com/prometheus/node_exporter/issues/1892) :
> 
>     1. `sudo apt install sysfsutils`
> 
>     2. Add this line in `/etc/sysfs.conf` : `mode class/powercap/intel-rapl:0/energy_uj = 0444`
> 
>     3. reboot
> 
> 
> Without rebooting you could do `sudo chmod -R a+r /sys/class/powercap/intel-rapl` but it will be lost at next boot.
> 
> But you need root access to do this. Following [#232](https://github.com/mlco2/codecarbon/issues/232) that has been closed by a PR that reduce errors logs, I create this issue to find a workaround that don't need to be root.
> 
> We could use the CPU load and apply it to the CPU TDP, but it is not so simple as shown here https://medium.com/teads-engineering/estimating-aws-ec2-instances-power-consumption-c9745e347959

**Comment by benoit-cty** [comment]
Hello,
You have to be root to do it. So try `sudo nano /etc/sysfs.conf` and add the line at the end of the file.

Or, if you don't want to use an editor :
```sh
echo 'mode class/powercap/intel-rapl:0/energy_uj = 0440'  | sudo tee -a /etc/sysfs.conf > /dev/null
echo 'owner class/powercap/intel-rapl:0/energy_uj = root:power'  | sudo tee -a /etc/sysfs.conf > /dev/null

```

**Comment by stefanovrizzi** [comment]
Thank you, I tried the two lines at the bottom, but I still get the following:

```
$ python
Python 3.12.3 (main, Feb  4 2025, 14:48:35) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from codecarbon import EmissionsTracker
>>> tracker = EmissionsTracker(force_cpu_power=45)
[codecarbon INFO @ 12:00:00] Codecarbon is taking the configuration from global file: /home/{username}/.codecarbon.config
[codecarbon INFO @ 12:00:00] [setup] RAM Tracking...
[codecarbon INFO @ 12:00:00] [setup] CPU Tracking...
[codecarbon WARNING @ 12:00:01] We saw that you have a {cpu type here} but we don't know it. Please contact us.
[codecarbon WARNING @ 12:00:01] We will use the default power consumption of 4 W per thread for your 16 CPU, so 64W.
[codecarbon INFO @ 12:00:01] Using user-provided CPU power: 45 Watts
[codecarbon WARNING @ 12:00:01] No CPU tracking mode found. Falling back on estimation based on TDP for CPU. 
 Linux OS detected: Please ensure RAPL files exist at /sys/class/powercap/intel-rapl/subsystem to measure CPU

[codecarbon INFO @ 12:00:01] CPU Model on constant consumption mode: {cpu type here}
[codecarbon WARNING @ 12:00:01] No CPU tracking mode found. Falling back on CPU load mode.
[codecarbon INFO @ 12:00:01] [setup] GPU Tracking...
[codecarbon INFO @ 12:00:02] Tracking Nvidia GPU via pynvml
[codecarbon INFO @ 12:00:02] The below tracking methods have been set up:
                RAM Tracking Method: RAM power estimation model
                CPU Tracking Method: cpu_load
                GPU Tracking Method: pynvml
            
[codecarbon INFO @ 12:00:02] >>> Tracker's metadata:
[codecarbon INFO @ 12:00:02]   Platform system: Linux-6.8.0-48-generic-x86_64-with-glibc2.39
[codecarbon INFO @ 12:00:02]   Python version: 3.12.3
[codecarbon INFO @ 12:00:02]   CodeCarbon version: 3.0.0_rc7
[codecarbon INFO @ 12:00:02]   Available RAM : 62.012 GB
[codecarbon INFO @ 12:00:02]   CPU count: 16 thread(s) in 1 physical CPU(s)
[codecarbon INFO @ 12:00:02]   CPU model: Intel(R) Xeon(R) W-11955M CPU @ 2.60GHz
[codecarbon INFO @ 12:00:02]   GPU count: 1
[codecarbon INFO @ 12:00:02]   GPU model: 1 x NVIDIA RTX A3000 Laptop GPU
[codecarbon INFO @ 12:00:05] Saving emissions data to file
```

I verified the existence of a RAPL file in the suggested path ```/sys/class/powercap/intel-rapl/subsystem$ ls```, I obtain:

```intel-rapl  intel-rapl:0  intel-rapl:0:0  intel-rapl:0:1  intel-rapl:1  intel-rapl-mmio  intel-rapl-mmio:0```

Is it normal?

Thank you

> Hello, You have to be root to do it. So try `sudo nano /etc/sysfs.conf` and add the line at the end of the file.
> 
> Or, if you don't want to use an editor :
> 
> echo 'mode class/powercap/intel-rapl:0/energy_uj = 0440'  | sudo tee -a /etc/sysfs.conf > /dev/null
> echo 'owner class/powercap/intel-rapl:0/energy_uj = root:power'  | sudo tee -a /etc/sysfs.conf > /dev/null

**Comment by benoit-cty** [comment]
Thanks for trying to debug this.

If you add `log_level="debug"` you will see a message "Not using the RAPL interface, an exception occurred while instantiating" with the reason of the fail.

**Comment by git997** [comment]
how can I use scaphandre in windows

**Comment by benoit-cty** [comment]
> how can I use scaphandre in windows

Hello @git997,
Here you are on CodeCarbon projet, for Scaphandre, see : https://hubblo-org.github.io/scaphandre-documentation/tutorials/installation-windows.html
